### PR TITLE
Implemented Gcloud function project in python for autosync branches. …

### LINKED ISSUE #12935: Automate branch synchronization
**Problem:**

All major branches (compatible/develop etc.) need to be in sync with master. Currently this is manual process of detecting changes between branches, creates PR, checks if there is no merge conflicts and push changes to target branch.
We would like to automate this process. 

**Requirements:**

- For MVP we want only merge conflicts detection
- There should NOT be a PR created if there is no merge conflicts
- There should be a PR created with assignee list (TBD) tagged. They should fix the PR
- Automation should detect changes immediately and perform merging attempt but once merge is failed there should not be one PR - one failed commit policy. There should be a solution to pileup failed commits and wrap it into single PR
- In future we maybe would like to attach some small buildkite pipeline for testing purposes

**Possible solution:**

For sure we need to utilize github api in order to detect changes and create PR. We did small research and there was one potential out of the box solution (https://github.com/wei/pull) but looks like it is not enough for out needs:
- it is more for fork syncing between repos
- doesn't have configuration in which PR won't be created on successful merge  etc.
- it lacks reference to buildkite

I performed spike in my private repo (i created it for testing purposes) and couldn't make it working at all.  I read that sometimes it's about couple of days to launch trigger as there are a lot of other repos that are using it and this app might be overloaded. 

Judging from above we think that custom solution should be proposed utlizing:
- Gitub REST API
- Buildkite REST API
- Google Cloud function 
- Programming language which has library for above

Initially we consider three technologies: [Rust / JS / Python] as those technology has low maintenance cost and are widely used in compay. After research we selected Python as rust doesn't have all required libraries and python is used a lot in automation scripts while js only in legacy solution that we would like to rewrite to python anyway.

As a result scope of this task would be:

**Configuration:**

- configure github webhook 
- determine set of source branches
- determine set of target branches
- obtain GITHUB and BUILDKITE tokens which necessary permissions

**FLOW**

1. Use Github webhook to launch script in google cloud function on new commit to specified branches
2. On event:
  - diff between source and target branch
   - create branch containing commits 
      a) If branch already exists, push this commit and tag assigners that there was yet another commit
  -  check if there are merge conflicts
      a) if there are: create pr from temp branch to target branch. Add proper description. Add assigners which should fix the pr
      b) if there are not: start buildkite pipeline (TBD) to verify changes. If they passes merge pr and exit
 
*Further remarks*

Knowing that meddling with automation process of github merging and pushing processes can be destructive  first script should be created on some different repo where we can perform cycle of errors->fixes before using it on mina repo.

### PR DESCRIPTION
**Pull request by dkijania**
Explain your changes:

Aim of this project is to satisfy needs for automatic branch synchronization between important branches in github. So far this is manual process of detecting changes between branches, Pull requests creation, checking if there is no merge conflicts and pushing changes to target branch.
This tool automate this process. It is possible to deploy project to google cloud function module.

Explain how you tested your changes:

I created side project 'tests' where i put e2e tests on some example repo (currently owned by me). Tool requires a lot of permissions so we need to be sure that it won't devastate out main repo

Checklist:

- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [x] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [x] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [x] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

* Closes #12935

### CODE REVIEW

File: automation/scripts/github_branch_autosync/LICENSE [hunk 1]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (psteckler) [inline automation/scripts/github_branch_autosync/LICENSE:1]: Why is an Apache license file needed here, there's already one in the repo root?

File: automation/scripts/github_branch_autosync/Makefile [hunk 2]
```diff
@@ -0,0 +1,48 @@
+clean:
+	find . -type f -name '*.py[co]' -delete -o -type d -name __pycache__ -delete
```
Comment (psteckler) [inline automation/scripts/github_branch_autosync/Makefile:2]: Using `find` to delete files feels a bit dangerous to me.

Would it be safer to delete files with `rm`, if their location is known?

**Comment by psteckler** [comment]
What kind of failures might occur when running this automation? If things go wrong, is recovery possible?

**Comment by dkijania** [comment]
> f this automation is introduced, it has to justify its complexity.
> Can you provide a quantifiable benefit of using this automation, over merging branches manually?

I think issue linked to this PR should explain why we need above algorithm (#12935). If it doesn't please let me know i will explain it a little bit deeper in linked issue

> Because this tool makes persistent changes to repos, it's important to know what actions it has performed.
> I see there are some print statements. Are these enough? Could we have logs with timestamps?
> Do you have the output from a test run?

This tool mainly should be used as cloud function. Based on logging recommendation from GCP, printing is enough [(docs)](https://cloud.google.com/functions/docs/monitoring/logging). Test run also should be as verbose as for main tool (if you think otherwise please speak out), and my goal was to present the test cases in a way that it is clear what should be an outcome of tool execution. However, I agree that maybe we should add dryrun mechanism, that won't do anything but just printout of operations it is meant to perform. Already added that in new commit.

> What kind of failures might occur when running this automation? If things go wrong, is recovery possible?

Tool performs operations such as:

- creates PR
- fast_forward branch
- merge branches
- creates temp branch
- removes temp branch

I think only fast_forwarding and merging is something that can disrupt repository (however there is still manual recovery in a form of resetting branches to former commits or reverting merge commits). Current version of tool does not run any buildkite pipeline to ensure correctness of merge. This is something that should be added in future.

**Comment by dkijania** [comment]
I noticed that some of the code were not pushed correctly as automatically generated .gitignore ignores lib folder :( Apologies for that.

**Comment by dkijania** [comment]
!ci-build-me

**Review by lk86 — APPROVED** [review]
I think the logic may need more work to handle all possible cases but that can be saved for a future PR. As designed, it looks great.

File: automation/scripts/github_branch_autosync/github_autosync/gcloud_entrypoint/main.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    if not source_branch in config.branches:
+        print(f"change in '{source_branch}' is not supported ")
+        return
+
+    target_branch = config.branches[source_branch]
+    github = GithubApi(config.github)
+    print(f"generating diff between {source_branch} and '{target_branch}'...")
+    cmp = github.get_diff_commits(target_branch,source_branch)
+
+    if cmp.status == "identical":
```
Comment (psteckler) [inline automation/scripts/github_branch_autosync/github_autosync/gcloud_entrypoint/main.py:40]: I think Python has a `match ... case` statement that might be useful here

File: automation/scripts/github_branch_autosync/Makefile (hunk 2 above)
Comment (psteckler) [reply automation/scripts/github_branch_autosync/Makefile:2]: ok, but "widely-used" doesn't mean "good" :-)

**Review by psteckler — APPROVED** [review]
if you can perform a dry run, that should alleviate most of my concerns

File: automation/scripts/github_branch_autosync/github_autosync/gcloud_entrypoint/main.py (hunk 3 above)
Comment (dkijania) [reply automation/scripts/github_branch_autosync/github_autosync/gcloud_entrypoint/main.py:40]: ok, i will fix that in next pr not to block dryrun testing

**Pull request merged**